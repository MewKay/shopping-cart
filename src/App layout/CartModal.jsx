import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import CartModalItem from "./CartModalItem";
import styles from "./CartModal.module.css";

const CartModal = ({
  showCartModal,
  totalQuantity,
  orderSubtotal,
  cart,
  handleHideCartModal,
}) => {
  const itemPlural = totalQuantity > 1 ? "s" : "";
  const modalOpenStyle = showCartModal ? styles["open"] : "";

  return (
    <div
      className={`
      ${styles["modal-overlay"]}
      ${modalOpenStyle}
      `}
      onClick={handleHideCartModal}
      aria-hidden={!showCartModal}
    >
      <div
        className={`
          ${styles["container"]}
          ${modalOpenStyle}
          `}
        onClick={(e) => e.stopPropagation()}
      >
        <button
          className={styles["close-modal-button"]}
          aria-label={"Close cart preview"}
          onClick={handleHideCartModal}
        >
          x
        </button>
        <div className={styles["cart-quantity-info"]}>
          Your cart contains {totalQuantity} item{itemPlural}
        </div>
        <div className={styles["cart-content"]}>
          <div className={styles["subtotal-info"]}>
            <p>Order Subtotal</p>
            <p className={styles["subtotal"]}>${orderSubtotal}</p>
          </div>
          <Link
            className={styles["cart-link"]}
            to={"/store/cart"}
            onClick={handleHideCartModal}
          >
            View or Edit Your Cart
          </Link>
          <ul className={styles["item-list"]}>
            {cart.map((item) => (
              <CartModalItem
                key={item.productDetails.id}
                image={item.productDetails.image}
                title={item.productDetails.title}
                price={item.productDetails.price}
                quantity={item.quantity}
              />
            ))}
          </ul>
        </div>
      </div>
    </div>
  );
};

CartModal.propTypes = {
  showCartModal: PropTypes.bool.isRequired,
  totalQuantity: PropTypes.number.isRequired,
  orderSubtotal: PropTypes.number.isRequired,
  cart: PropTypes.arrayOf(
    PropTypes.shape({
      productDetails: PropTypes.shape({
        id: PropTypes.number.isRequired,
        image: PropTypes.string,
        title: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
      }),
      quantity: PropTypes.number.isRequired,
    })
  ),
  handleHideCartModal: PropTypes.func.isRequired,
};

export default CartModal;
