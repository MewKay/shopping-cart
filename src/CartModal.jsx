import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import CartModalItem from "./CartModalItem";

const CartModal = ({
  totalQuantity,
  orderSubtotal,
  cart,
  handleHideCartModal,
}) => {
  const itemPlural = totalQuantity > 1 && "s";
  return (
    <div>
      <div>
        <button aria-label={"Close cart preview"} onClick={handleHideCartModal}>
          x
        </button>
        <div>
          Your cart contains {totalQuantity} item{itemPlural}
        </div>
        <div>
          <div>
            <p>Order Subtotal</p>
            <p>${orderSubtotal}</p>
          </div>
          <Link to={"/store/cart"}>View or Edit Your Cart</Link>
          <ul>
            {cart.map((item) => (
              <CartModalItem
                key={item.productDetails.id}
                image={item.productDetails.image}
                title={item.productDetails.title}
                price={item.productDetails.price}
                quantity={item.quantity}
              />
            ))}
          </ul>
        </div>
      </div>
    </div>
  );
};

CartModal.propTypes = {
  totalQuantity: PropTypes.number.isRequired,
  orderSubtotal: PropTypes.number.isRequired,
  cart: PropTypes.arrayOf(
    PropTypes.shape({
      productDetails: PropTypes.shape({
        id: PropTypes.number.isRequired,
        image: PropTypes.string,
        title: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
      }),
      quantity: PropTypes.number.isRequired,
    })
  ),
  handleHideCartModal: PropTypes.func.isRequired,
};

export default CartModal;
