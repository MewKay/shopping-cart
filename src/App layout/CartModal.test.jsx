import { describe, it, expect, vi } from "vitest";
import userEvent from "@testing-library/user-event";
import { screen, render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import CartModal from "./CartModal";

vi.mock("./CartModalItem", () => ({
  default: ({ image, title, price, quantity }) => (
    <li>
      <img src={image} alt="" />
      <p>{title}</p>
      <p>{price}</p>
      <p>{quantity}</p>
    </li>
  ),
}));

//Component props
const showCartModal = true;
const totalQuantity = 10;
const orderSubtotal = 352.45;
const cart = [
  {
    productDetails: {
      id: 0,
      image: "https://image.placeholder.com/product",
      title: "Product 0",
      price: 49.5,
    },
    quantity: 5,
  },
  {
    productDetails: {
      id: 1,
      image: "https://image.placeholder.com/product",
      title: "Product 1",
      price: 20.99,
    },
    quantity: 5,
  },
];
const handleHideCartModal = vi.fn();

describe("CartModal component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <CartModal
        showCartModal={showCartModal}
        totalQuantity={totalQuantity}
        orderSubtotal={orderSubtotal}
        cart={cart}
        handleHideCartModal={handleHideCartModal}
      />,
      { wrapper: MemoryRouter }
    );

    expect(container).toMatchSnapshot();
  });

  it("calls 'handleHideCartModal' when 'Close cart preview' button is clicked", async () => {
    const user = userEvent.setup();
    render(
      <CartModal
        showCartModal={showCartModal}
        totalQuantity={totalQuantity}
        orderSubtotal={orderSubtotal}
        cart={cart}
        handleHideCartModal={handleHideCartModal}
      />,
      { wrapper: MemoryRouter }
    );

    const closeButton = screen.getByLabelText("Close cart preview");
    await user.click(closeButton);
    expect(handleHideCartModal).toHaveBeenCalledOnce();
  });

  it("adds plural on 'item' text if more than 1", () => {
    render(
      <CartModal
        showCartModal={showCartModal}
        totalQuantity={totalQuantity}
        orderSubtotal={orderSubtotal}
        cart={cart}
        handleHideCartModal={handleHideCartModal}
      />,
      { wrapper: MemoryRouter }
    );

    const itemText = screen.getByText(/items/);

    expect(itemText).toBeInTheDocument();
  });

  it("does not add plural on 'item' text if 1 or less", () => {
    const mockTotalQuantity = 1;
    render(
      <CartModal
        showCartModal={showCartModal}
        totalQuantity={mockTotalQuantity}
        orderSubtotal={orderSubtotal}
        cart={cart}
        handleHideCartModal={handleHideCartModal}
      />,
      { wrapper: MemoryRouter }
    );

    const itemText = screen.getByText(/item(?= |$)/);

    expect(itemText).toBeInTheDocument();
  });
});
