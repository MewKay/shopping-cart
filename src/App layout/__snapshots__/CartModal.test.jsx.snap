// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartModal component > renders correctly 1`] = `
<div>
  <div
    aria-hidden="false"
    class="
      _modal-overlay_8d89e9
      _open_8d89e9
      "
  >
    <div
      class="
          _container_8d89e9
          _open_8d89e9
          "
    >
      <button
        aria-label="Close cart preview"
        class="_close-modal-button_8d89e9"
      >
        <svg
          class="lucide lucide-x"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
      <div
        class="_cart-quantity-info_8d89e9"
      >
        Your cart contains 
        10
         item
        s
      </div>
      <div
        class="_cart-content_8d89e9"
      >
        <div
          class="_subtotal-info_8d89e9"
        >
          <p>
            Order Subtotal
          </p>
          <p
            class="_subtotal_8d89e9"
          >
            $
            352.45
          </p>
        </div>
        <a
          class="_cart-link_8d89e9"
          href="/store/cart"
        >
          View or Edit Your Cart
        </a>
        <ul
          class="_item-list_8d89e9"
        >
          <li>
            <img
              alt=""
              src="https://image.placeholder.com/product"
            />
            <p>
              Product 0
            </p>
            <p>
              49.5
            </p>
            <p>
              5
            </p>
          </li>
          <li>
            <img
              alt=""
              src="https://image.placeholder.com/product"
            />
            <p>
              Product 1
            </p>
            <p>
              20.99
            </p>
            <p>
              5
            </p>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
